### ISSUE TITLE
Re-examine allowing deferred type inference

### ISSUE THREAD

**Issue by cconvey** [NervanaSystems/ngraph#82]
The current (as of this writing) NGraph C++ API does not propagate type information incrementally, as each node is added to a graph.  Instead, there's a mechanism for propagating it at a later date.

Two reasons were given for this design choice:
1. In the future, we may need to support frameworks that provide type information only after revealing other aspects of a graph's topology.
2. We may want to periodically re-propagate or re-check type information due to graph transformations.

A brief Slack discussion is found [here](https://nervana.slack.com/archives/C5SD44PL1/p1504642220000031).

For this Issue, re-examine the reasoning for this design choice, and consider the merits of simply propagating type information incrementally each time a node is added to the graph.

Attention: @diyessi @aprocter @cconvey.
